packages/clay-date-picker/src/ClayDatePicker.tsx

```diff
@@ -0,0 +1,379 @@
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import moment from 'moment';
+import React, {
+	ChangeEvent,
+	FunctionComponent,
+	ReactNode,
+	useEffect,
+	useRef,
+	useState,
+} from 'react';
+
+import {useCurrentMonth, useCurrentTime, useWeeks} from './Hooks';
+
+import Button from './Button';
+import DateNavigation from './DateNavigation';
+import DayNumber from './DayNumber';
+import DaysTable from './DaysTable';
+import Dropdown from './Dropdown';
+import InputDate from './InputDate';
+import TimePicker from './TimePicker';
+import Weekday from './Weekday';
+import WeekdayHeader from './WeekdayHeader';
+
+export enum firstDayOfWeek {
+	Sunday = 0,
+	Monday = 1,
+	Tuesday = 2,
+	Wednesday = 3,
+	Thursday = 4,
+	Friday = 5,
+	Saturday = 6,
+}
+
+export interface IAriaLabels {
+	buttonDot: string;
+	buttonNextMonth: string;
+	buttonPreviousMonth: string;
+	input?: string;
+}
+
+export interface IYears {
+	end: number;
+	start: number;
+}
+
+interface Props {
+	/**
+	 * Labels for the aria attributes
+	 */
+	ariaLabels?: IAriaLabels;
+
+	/**
+	 * Set the format of how the date will appear in the input element.
+	 * See available: https://momentjs.com/docs/#/parsing/string-format/
+	 */
+	dateFormat?: string;
+
+	/**
+	 * Set the first day of the week, starting from
+	 * 0 (Sunday) to 6 (Saturday).
+	 */
+	firstDayOfWeek?: firstDayOfWeek;
+
+	/**
+	 * Function that should return the React element to
+	 * render on the datepicker footer.
+	 */
+
+	footerElement?: (object: {spritemap: string}) => ReactNode;
+
+	/**
+	 * Id to be applied to the element.
+	 */
+	id?: string;
+
+	/**
+	 * Flag to indicate if date is expanded.
+	 */
+	initialExpanded?: boolean;
+
+	/**
+	 * The month to display in the calendar on the first render.
+	 */
+	initialMonth?: Date;
+
+	/**
+	 * Name of the input.
+	 */
+	inputName?: string;
+
+	/**
+	 * The names of the months.
+	 */
+	months?: Array<string>;
+
+	/**
+	 * Called when the input change.
+	 */
+	onValueChange: (value: Date | string) => void;
+
+	/**
+	 * Describe a brief tip to help users interact.
+	 */
+	placeholder?: string;
+
+	/**
+	 * The path to the SVG spritemap file containing the icons.
+	 */
+	spritemap: string;
+
+	/**
+	 * Flag to enable datetime selection.
+	 */
+	time?: boolean;
+
+	/**
+	 * Set the format of how the time will appear in the input element.
+	 * See available: https://momentjs.com/docs/#/parsing/string-format/
+	 */
+	timeFormat?: string;
+
+	/**
+	 * Flag to indicate the timezone of the Time Picker.
+	 */
+	timezone?: string;
+
+	/**
+	 * Flag to indicate whether to use native date picker
+	 */
+	useNative?: boolean;
+
+	/**
+	 * Set the value of the input.
+	 */
+	value: string;
+
+	/**
+	 * Short names of days of the week to use in the header
+	 * of the month. It should start from Sunday.
+	 */
+	weekdaysShort?: Array<string>;
+
+	/**
+	 * List of years available for navigation within the selector.
+	 */
+	years?: IYears;
+}
+
+const DateNow = new Date();
```

Comment:
Well thinking about this, maybe we can always make sure to clone `initialMonth` before sending to `onValueChange` https://github.com/liferay/clay/pull/1440/files?file-filters%5B%5D=.tsx#diff-ff7375578ff0fc5080ff0275edb83851R262

Reply:
Here's another one: I open portal at 11:45 PM on December 31st, the navigate another page with a date picker at 12:15 AM. Because my system administrator kindly configured a long timeout, I wasn't logged out due to inactivity while I celebrated the arrival of the new year, and this new date picker now starts off showing the previous month.

(I know, I know... but it is late on Friday and my mind is wandering.) 😂

Reply:
Yeah, this is a very particular ​​problem 😅, so I think we should set an `initialMonth` whenever we have a call to `setExpanded` if it is not passed. It's the best place I think to add that. what do you think?

Reply:
The most natural change would be switching `initialMonth = DateNow,` to `initialMonth = new Date(),`, right? I don't know if it will work, because `useCurrentMonth` is a little confusing to me now that I look at it.

If I'm reading `useCurrentMonth` correctly, the value we pass for `initialMonth` is going to be used in a callback that is passed as the initial state in a `useState()` call — question: why aren't we setting the result of calling `setDate()` as the initial value? Instead, we pass a callback that sets it.

That is, why is it?

```javascript
const useCurrentMonth = (value: Date) => {
	const setDate = (date: Date) =>
		moment(date)
			.clone()
			.set('date', 1)
			.set('hour', 12)
			.set('minute', 0)
			.set('second', 0)
			.set('millisecond', 0)
			.toDate();

	const [currentMonth, set] = useState<Date>(() => setDate(value));

	const setCurrentMonth = (value: Date) => {
		set(setDate(value));
	};

	return [currentMonth, setCurrentMonth] as [Date, (value: Date) => void];
};
```

instead of:

```javascript
const useCurrentMonth = (value: Date) => {
	const setDate = (date: Date) =>
		moment(date)
			.clone()
			.set('date', 1)
			.set('hour', 12)
			.set('minute', 0)
			.set('second', 0)
			.set('millisecond', 0)
			.toDate();

	const [currentMonth, set] = useState<Date>(setDate(value));

	const setCurrentMonth = (value: Date) => set(setDate(value));

	return [currentMonth, setCurrentMonth];
};
```

(I am just guessing, but I get the impression that you might have needed that `as` cast only because of passing the function instead of the value.)

Reply:
I think even changing `initialMonth = new Date()` might not take effect in some cases, because we are only initializing the `currentMonth` state only when the component is created. We would have to make sure to update the `initialMonth` even when it is already mount, we would have a case of the person being midnight at work, but a few minutes before he left the page with the Date Picker open, with the 31/03 at midnight he back to the page, but there still is with the 31/03, in case it should be 01/04.

About `useCurrentMonth` I ended up removing it, but it reflects on the other Hooks we have. The use of `as` is to infer the return of the function, even if we do something like:
```javascript
const setCurrentMonth = (value: Date) => set(setDate(value));
```

TypeScript is not inferring the return from `set` to `setCurrentMonth`, so we have an inference of the wrong kind.
```
IDay[][] | ((value: Date) => void)
```

Reply:
As a user, though, I don't expect a mounted component that is already on the page to flip months at midnight; what I do expect, however, is that if I navigate to a new view and see a date picker, on the new view it will always start on the current month.

Reply:
Yes the user would not wait so long with an open page. But it was a case that a person is with the page open a few minutes before midnight, although it is a case that may be very difficult to happen.

Reply:
Yeah, but I mean independently of whether I have the page open 5 seconds/minutes or 5 days/weeks, I don't expect a rendered date picker to start rendering a different month without me explicitly skipping to another month. (On the other hand, if I navigate to a new view, I'd expect a new date picker to start on whatever month is the current month.)

Reply:
Oh yeah, you're right.